feat: enable key rotation

sreya opened the pull request:
This PR contains the remaining logic necessary to hook up key rotation to the product.

closes https://github.com/coder/internal/issues/52

sreya committed:
feat: enable key rotation

sreya committed:
add migration

sreya committed:
Refactor cryptographic key handling for OIDC and API keys

This commit updates the cryptographic key handling by separating
workspace app token and API key features. It corrects feature identifiers
for clearer distinction between OIDC conversion and API key usage, enhancing
the maintainability and clarity around cryptographic key usage within the
system. Additionally, reworks related tests and migration scripts to align
with these changes.

sreya committed:
the end is nigh

sreya committed:
fix migrations

sreya committed:
Refactor key cache management for better clarity

- Consolidates key cache handling by replacing legacy key cache references with central key caches.
- Enhances modularity and maintainability by using consistent key management methods.
- Removes redundant `StaticKeyManager` implementation for streamlined code.
- Adjusts cryptographic key generation and cache utilization across critical components.

sreya committed:
hm

sreya committed:
fixing tests

sreya committed:
time to fix it

sreya committed:
Refactor cryptokeys Fetcher to include feature param

Enhances flexibility by making the `Fetcher` interface
receive a `CryptoKeyFeature` parameter. This change aligns
various call sites that implement or utilize `Fetcher`, allowing
for more granular queries.

sreya committed:
Refactor key caching and logging behavior

- Improve code clarity by refactoring key caching logic.
- Simplify logger initialization for signing and encryption caches.
- Ensure consistent closing of caches in the API and server.

sreya committed:
Refactor crypto_key_feature migration logic

This change enhances the crypto_keys table by updating enum
handling for features. It introduces distinct roles for key
storage, aiding in better distinction and maintenance of key
features.

sreya committed:
Refactor key management and enhance logging

- Improve clarity by naming loggers used in key cache creation.
- Adjust key cache context to utilize KeyReader for consistent
  context handling.
- Refactor API to use central key cache management approach.
- Enhance error messages for crypto key fetching.
- Update test to add safety against unexpected panics.

sreya committed:
Update cryptokey feature test and migration logic

- Enhance testing scenarios for cryptokey features including cases
  with no keys and specific key states.
- Update tests to ensure new cryptokey features are handled correctly.
- Remove outdated migration scripts for cryptokey features as they
  are not required anymore.
- Refactor workspace proxy keys to only allow whitelisted
  cryptokey features, improving security and stability.

sreya committed:
gen

sreya committed:
Refactor cryptokeys cache to include key reader context

Enhances the cache functionality by wrapping the context with a key
reader, ensuring proper authorization checks during cryptographic
operations. This change aligns cache behavior with security practices.

sreya committed:
Refactor jwtutils to remove redundant key reader

sreya committed:
Remove unused cryptokey feature fixtures

sreya committed:
Refactor cryptokeys comments and variable typo

Adjust comments to accurately describe cache functionality
and fix a typo in the `New` function parameters for clarity.

sreya committed:
fix comments

sreya committed:
move rotator out of coderd

sreya committed:
remove composite jwtutil interfaces

sreya committed:
fix tests caused by moving rotator initiation out of coderd

spikecurtis commented on coderd/workspaceapps.go:
```diff
@@ -122,8 +123,7 @@ func (api *API) workspaceApplicationAuth(rw http.ResponseWriter, r *http.Request
 
-	// Encrypt the API key.
-	encryptedAPIKey, err := api.AppSecurityKey.EncryptAPIKey(workspaceapps.EncryptedAPIKeyPayload{
+	encryptedAPIKey, err := jwtutils.Encrypt(ctx, api.AppEncryptionKeyCache, workspaceapps.EncryptedAPIKeyPayload{
```
You're missing the 1 minute expiration.

While we're at it, we should set a NotBefore to 1 minute in the past (to allow for minor clock skew between coderd and the proxy), an issuer (coderd) and an audience (wsproxy), and check all of these on decryption.

sreya replied:
🤦 good catch

spikecurtis commented on coderd/workspaceapps/proxy.go:
```diff
@@ -176,7 +180,10 @@ func (s *Server) handleAPIKeySmuggling(rw http.ResponseWriter, r *http.Request,
+	var payload EncryptedAPIKeyPayload
+	err := jwtutils.Decrypt(ctx, s.APIKeyEncryptionKeycache, encryptedAPIKey, &payload, jwtutils.WithDecryptExpected(jwt.Expected{
+		Time: time.Now(),
+	}))
```
If we get an error here, we redirect to a static error page, which is bad news if a user gets caught with an unsupported JWE without a key ID.  

Maybe we should detect a missing key ID and just strip the query parameter?

I guess you'd have to be pretty unlucky to be doing this API key dance _right_ as Coder is being upgraded. So, maybe it's not worth doing. WDYT?

sreya replied:
Hmm yeah, pretty unlikely and will only result in a one-off error right?

spikecurtis replied:
I think so

spikecurtis commented on enterprise/workspaceapps_test.go:
```diff
@@ -59,6 +67,13 @@ func TestWorkspaceApps(t *testing.T) {
+		})
+		_ = dbgen.CryptoKey(t, db, database.CryptoKey{
+			Feature: database.CryptoKeyFeatureWorkspaceAppsAPIKey,
+		})
```
This strikes me as an argument for making the key rotator part of the `API`, or at least part of `coderdtest` / `coderdenttest`.  We treat these as being a "fully functioning" coderd, and without crypto keys, certain parts are not functional.

Other test writers are going to get bit with not having the crypto keys and will be confused.

sreya replied:
Yeah I actually had it this way and moved it out but I think I regret that. I think we move it into the API.

spikecurtis commented on tailnet/resume_test.go:
```diff
@@ -121,61 +24,76 @@ func TestResumeTokenKeyProvider(t *testing.T) {
+
+func newKeySigner(key tailnet.ResumeTokenSigningKey) cryptokeys.StaticKey {
+	return cryptokeys.StaticKey{
+		ID:  uuid.New().String(),
```
I know these are just strings, but for consistency, we should use integers not UUID to match what we actually use in production.

spikecurtis commented on codersdk/workspacesdk/connector_internal_test.go:
```diff
@@ -165,13 +166,17 @@ func TestTailnetAPIConnector_ResumeToken(t *testing.T) {
 	require.NoError(t, err)
-	resumeTokenProvider := tailnet.NewResumeTokenKeyProvider(resumeTokenSigningKey, clock, time.Hour)
+	mgr := cryptokeys.StaticKey{
+		ID:  uuid.New().String(),
```
use integer ID for consistency

sreya commented on coderd/coderd.go:
```diff
@@ -444,6 +448,38 @@ func New(options *Options) *API {
+			fetcher,
+			codersdk.CryptoKeyFeatureOIDCConvert,
+		)
+		must(options.Logger, "start oidc convert key cache", err)
```
Why not? I can't think of a good reason why trying to query keys should fail which means there's probably something wrong with our DB connection.

spikecurtis replied:
In distributed systems you want to minimize coupling and self-heal.  A particularly annoying sort of coupling is the order in which subcomponents start up.  That is, you want it to be the case that things work correctly if coderd and the database both are up, regardless of whether coderd started before the database. Ordering dependencies can be a nightmare as you add more systems. It might not just be a database server: you might also need the DNS route to have been propagated, or an authenticating proxy to be up, etc.

spikecurtis replied:
I realize there are other things that will fail if the DB is unreachable, but let's not make our problems worse.

sreya commented on coderd/database/migrations/000266_cryptokey_features.up.sql:
```diff
@@ -0,0 +1,18 @@
+
+DROP TYPE crypto_key_feature;
+
+ALTER TYPE new_crypto_key_feature RENAME TO crypto_key_feature;
```
Is this a separate issue from [this comment](https://github.com/coder/coder/pull/15066#discussion_r1804521384). The only time redirecting would be helpful is if you got caught mid-upgrade right? If you're in a scenario where you're partially upgraded (say the `wsproxy` is using old-style and `coderd` is using new-style) redirecting to reauth won't help and in fact could make it more confusing because you're going to hit `too many redirects`

sreya replied:
So just to summarize the consequences:
- `oidc_convert`: Will throw an error asking the user to retry. This should be the rarest of them all and should be a one-off error
- `workspace_apps_api_key`: Will throw an error asking the user to retry. Should be the second rarest since most users will already be authenticated.
- `workspace_apps_token`: Will silently discard token and issue a new token (presuming the api key is still valid)
- `tailnet_resume`: Will ignore the token and generate a new peer id

I've added tests for each case thus the balloon in PR size 😞

sreya committed:
pr comments

sreya committed:
Merge branch 'main' into jon/glue

sreya committed:
Refactor tests to remove direct database setup

Refactored various tests to eliminate the need for manually setting up
a database connection and generating crypto keys. This change simplifies
the test setup and makes it less dependent on database-specific
configurations, aligning with recent structural refactoring.

sreya committed:
Rename cryptokey migration files to update sequence

sreya committed:
Fix conditional logging in key cache initialization

Ensure that logging only occurs when an error is present during the initialization of the key caches. This prevents unnecessary log entries when no error is encountered, improving log clarity and reducing noise.

sreya committed:
Refactor crypto key management in tests

Remove redundant database setup for tests by generating keys in memory.
Enhance tests by validating both inclusion and exclusion of specific keys.

sreya committed:
add test for oidc jwt

sreya committed:
add test for tailnet_resume jwt

sreya committed:
add test for workspaceapps

sreya committed:
add test for signedtoken

sreya committed:
fix migrations

sreya committed:
fmt

spikecurtis commented on coderd/coderd.go:
```diff
@@ -444,6 +451,52 @@ func New(options *Options) *API {
+
+	// Start a background process that rotates keys. We intentionally start this after the caches
+	// are created to force initial requests for a key to populate the caches. This helps catch
+	// bugs that may only occur when a key isn't precached in tests and the latency cost is minimal.
```
👍

spikecurtis commented on coderd/coderd.go:
```diff
@@ -444,6 +451,52 @@ func New(options *Options) *API {
+	// bugs that may only occur when a key isn't precached in tests and the latency cost is minimal.
+	err = cryptokeys.StartRotator(ctx, options.Logger, options.Database)
+	if err != nil {
+		must(options.Logger, "failed to start key rotator", err)
```
This is in some ways even worse than failing on the caches, since we try to rotate in a repeatable read transaction without retries.  If you start multiple coderds simultaneously, they could easily interfere, and some coderds will crash, even if the database is up and working perfectly!

spikecurtis commented on coderd/coderd.go:
```diff
@@ -606,7 +659,7 @@ func New(options *Options) *API {
 	})
 	if err != nil {
-		api.Logger.Fatal(api.ctx, "failed to initialize tailnet client service", slog.Error(err))
+		must(api.Logger, "failed to initialize tailnet client service", err)
```
you're checking for a nil error which `must` also does, so I don't think `must` adds much clarity if you use it like this.

spikecurtis commented on coderd/workspaceapps_test.go:
```diff
@@ -240,7 +259,15 @@ func TestWorkspaceApplicationAuth(t *testing.T) {
+			}))
+			require.NoError(t, err)
+			require.Equal(t, jwt.NewNumericDate(clock.Now().Add(time.Minute)), token.Expiry)
+			require.Equal(t, jwt.NewNumericDate(clock.Now().Add(-time.Minute)), token.NotBefore)
```
Niice.

sreya commented on coderd/cryptokeys/cache.go:
```diff
@@ -134,12 +136,13 @@ func newCache(ctx context.Context, logger slog.Logger, fetcher Fetcher, feature
 	if err != nil {
-		cache.refreshCancel()
+		cache.cancel()
 		return nil, xerrors.Errorf("initial fetch: %w", err)
```
tbf it's a fatal error so it would cause the entire process to crash. But yeah we can treat it like the caches.

sreya commented on coderd/workspaceapps/apptest/apptest.go:
```diff
@@ -580,6 +642,69 @@ func Run(t *testing.T, appHostIsPrimary bool, factory DeploymentFactory) {
+					require.Equal(t, u.String(), gotLocation.Query().Get("redirect_uri"))
+
+					// Load the application auth-redirect endpoint.
+					resp, err = requestWithRetries(ctx, t, appDetails.SDKClient, http.MethodGet, "/api/v2/applications/auth-redirect", nil, codersdk.WithQueryParam(
```
Yeah that's fair

sreya committed:
pr comments

sreya committed:
Merge branch 'main' into jon/glue

sreya committed:
Rename cryptokey migration files to update sequence

spikecurtis reviewed (approved):
LGTM

sreya committed:
Merge branch 'main' into jon/glue

sreya committed:
migrations

sreya committed:
Refactor StaticKey to jwtutils package

移動StaticKeyからcryptokeysパッケージへのjwtutils。
これにより、jwtutilsとのセキュリティおよび独立性の一貫性が強化されます。
また、tailnetの依存関係を減らします。

sreya committed:
fix tests
